### PR TITLE
Merge mainline commits as of 2026/02/08 05:16 UTC with MPS optimizations and do optimization checks

### PR DESCRIPTION
**Pull request by riversedge**
This is an attempt to take the > 100 commits from the past few days and have AI merge them in with other MPS related optimizations.   Anytime there is a conflict was directed to choose the more optimal path.   After commit took a pass looking for and bugs or suboptimal areas and addressed with changes.   

## Summary by CodeRabbit

* **New Features**
  * Added `--allowed-path` CLI flag to specify custom Gradio-allowed file paths.
  * Job progress tracking with stage information (queued, running, succeeded, failed).
  * Resume training from checkpoint directory.
  * MPS memory detection with MAX_MPS_VRAM environment variable support.
  * Audio duration inference for GPU constraint clamping.

* **Bug Fixes**
  * Safety fallback converts cover tasks to text2music when source audio is missing.
  * Improved device fallback handling for unavailable GPUs/accelerators.

* **Enhancements**
  * Device-aware mixed precision and dataloader configuration.
  * Throttled progress callbacks to reduce cache writes.
  * Added memory-aware model initialization and anti-clipping normalization.

* **Dependencies**
  * Added tensorboard>=2.20.0.

### CODE REVIEW

**Review by ChuxiJ — APPROVED** [review]

**Comment by djtuBIG-MaliceX** [comment]
This merge has broken audio generations for M3 Ultra on macOS sequoia.  Reverting to c3dcf14 makes it work again.

No settings are able to successfully create an mp3 since commit 17825ee. evidence result in the following:

image

File: acestep/gradio_ui/events/generation_handlers.py [hunk 1]
```diff
@@ -462,7 +462,7 @@ def init_service_wrapper(dit_handler, llm_handler, checkpoint, config_path, devi
             backend=backend,
             device=device,
             offload_to_cpu=offload_to_cpu,
-            dtype=dit_handler.dtype,
+            dtype=None,
```
Comment (djtuBIG-MaliceX) [inline acestep/gradio_ui/events/generation_handlers.py:465]: Why is this None?

File: acestep/acestep_v15_pipeline.py [hunk 2]
```diff
@@ -281,7 +287,7 @@ def main():
                         backend=args.backend,
                         device=args.device,
                         offload_to_cpu=args.offload_to_cpu,
-                        dtype=dit_handler.dtype,
+                        dtype=None,
```
Comment (djtuBIG-MaliceX) [inline acestep/acestep_v15_pipeline.py:290]: Why is this None?

File: acestep/llm_inference.py [hunk 3]
```diff
@@ -2897,7 +2977,9 @@ def _load_model_context(self):
             # Clear accelerator cache after offloading
             if torch.cuda.is_available():
                 torch.cuda.empty_cache()
-            elif hasattr(torch.backends, "mps") and torch.backends.mps.is_available():
+            elif hasattr(torch, 'xpu') and torch.xpu.is_available():
+                torch.xpu.empty_cache()
+            elif hasattr(torch.backends, "mps") and torch.backends.mps.is_available() and hasattr(torch, "mps") and hasattr(torch.mps, "empty_cache"):
```
Comment (djtuBIG-MaliceX) [inline acestep/llm_inference.py:2982]: You don't need to check for empty_cache(), already defined 3y ago per  https://github.com/pytorch/pytorch/blame/449b1768410104d3ed79d3bcfe4ba1d65c7f22c0/torch/mps/__init__.py#L82

File: acestep/llm_inference.py [hunk 4]
```diff
@@ -72,6 +73,41 @@ def __init__(self, persistent_storage_path: Optional[str] = None):
...
+            except Exception:
+                pass
+            if torch.cuda.is_available():
+                torch.cuda.empty_cache()
+                torch.cuda.synchronize()
+            elif hasattr(torch, "mps") and torch.backends.mps.is_available():
+                if hasattr(torch.mps, "synchronize"):
+                    torch.mps.synchronize()
+                if hasattr(torch.mps, "empty_cache"):
+                    torch.mps.empty_cache()
```
Comment (djtuBIG-MaliceX) [inline acestep/llm_inference.py:104]: Don't need to check for these. If you do, then you should have an else to log a warning.

File: acestep/inference.py [hunk 5]
```diff
@@ -383,11 +397,95 @@ def generate_music(
...
+                                # Example: "duration:  183.165s"
+                                parts = line.strip().split()
+                                for p in parts:
+                                    if p.endswith("s"):
+                                        try:
+                                            return float(p.rstrip("s"))
+                                        except ValueError:
+                                            continue
+                except Exception:
+                    pass
```
Comment (djtuBIG-MaliceX) [inline acestep/inference.py:447]: Should be logging errors.

File: acestep/gradio_ui/events/training_handlers.py [hunk 6]
```diff
@@ -603,6 +604,48 @@ def start_training(
...
+            mixed_precision = "bf16"
+        elif device_type == "xpu":
+            num_workers = 4
+            pin_memory = True
+            prefetch_factor = 2
+            persistent_workers = True
+            pin_memory_device = None
+            mixed_precision = "bf16"
+        elif device_type == "mps":
+            num_workers = 0
```
Comment (djtuBIG-MaliceX) [inline acestep/gradio_ui/events/training_handlers.py:629]: Why is this 0?

File: acestep/gradio_ui/events/training_handlers.py [hunk 7]
```diff
@@ -603,6 +604,48 @@ def start_training(
...
+            pin_memory = False
+            prefetch_factor = 2
+            persistent_workers = False
+            pin_memory_device = None
+            mixed_precision = "fp16"
```
Comment (djtuBIG-MaliceX) [inline acestep/gradio_ui/events/training_handlers.py:634]: Why is it set to fp16 for mps when you say to use fp32 in https://github.com/ace-step/ACE-Step-1.5/pull/322/changes#diff-40742911b92dacba8e2575fd9cc167029f534ddc7a4079b8757d9c621c0ed70fR452 ?

File: openrouter/openrouter_api_server.py [hunk 8]
```diff
@@ -567,7 +567,7 @@ async def lifespan(app: FastAPI):
                 backend=backend,
                 device=device,
                 offload_to_cpu=lm_offload,
-                dtype=handler.dtype,
+                dtype=None,
```
Comment (djtuBIG-MaliceX) [inline openrouter/openrouter_api_server.py:570]: Why is this None?

File: cli.py [hunk 9]
```diff
@@ -1398,7 +1398,7 @@ def main():
             backend=args.backend,
             device=device,
             offload_to_cpu=args.offload_to_cpu,
-            dtype=dit_handler.dtype,
+            dtype=None,
```
Comment (djtuBIG-MaliceX) [inline cli.py:1401]: Why is this None?
